Port to C++11/C++14 removing superflous Boost facilities

Flamefire opened the pull request:
This addresses #69 by replacing Boost components for which equivalents or improved versions in C++11/14 exist.

This will reduce compile times for real world projects and potentially make test easier to debug.

Flamefire commented:
@mat007 I'm done for the first step, appveyor failure is doc generation related and just a HTTP error.

I'll rebase once #90 is merged to make this a bit smaller but you can have a look already. Next step would be replacing Boost.PP usage by variadic templates which should make the code much more understandable. If wanted I can try to port the inclusion of the benchmarks into the build to current master so we can compare build times before/after for those. Note that those are not fully accurate as one can assume the std headers are used already while the benchmarks assume a "clean state"

Flamefire commented:
@mat007 Ready. The decreased coverage is due to https://coveralls.io/builds/32087216/source?filename=include/turtle/detail/expectation_template.hpp where I reduced the number of lines by 7 which causes a decrease in relative coverage. Might be worth checking if we can test the currently uncovered serialization functions.

mat007 commented on doc/example/getting_started.cpp:
```diff
@@ -112,18 +167,49 @@ MOCK_BASE_CLASS( mock_view, view )
    calculator c( v );
    MOCK_EXPECT( v.display ).once().with( 0 ); // missing returns( true )
    c.add( 0, 0 );
+   assert_failure("missing action");
```
This is showing up in the docs, I don’t think this was intentional, was it?

Flamefire replied:
It is. The docs say that at this point a test assertion fails that an action is missing. This highlights it AND actually tests it

mat007 replied:
I’m thinking some newcomers will be puzzled by this line, especially since the definition for `assert_failure` will not be visible.
There should maybe at least be a comment?

Flamefire replied:
If we upgrade the required boost version to 1.59 (or higher) we can use assertions in fixtures teardown

- Initial support via `BOOST_AUTO_TEST_CASE(test, *utf::fixture<MixFixture>)` is in 1.59
- Better in 1.65 via  `BOOST_FIXTURE_TEST_CASE(test, MyFixture)` (like now but teardown() is supported in 1.65+)

What shall it be?

mat007 replied:
I think 1.65 is fine if that helps.

Flamefire replied:
Upgraded that for new CI

mat007 commented on doc/example/getting_started.cpp:
```diff
@@ -112,18 +167,49 @@ MOCK_BASE_CLASS( mock_view, view )
+    view& v;
 public:
-    calculator( view& v );
+    calculator( view& v ): v(v) {}
+
+    void add( int a, int b ){ v.display(a + b); }
```
Code formatting?

mat007 commented on doc/example/getting_started.cpp:
```diff
@@ -112,18 +167,49 @@ MOCK_BASE_CLASS( mock_view, view )
+    void add( int a, int b ){ v.display(a + b); }
+};
+
+struct CatchFailureFixture
```
Why do we need a custom error policy?

Flamefire replied:
To test for failure. We want the test to pass when the mock reports a test failure. In boost test there is no way to check if the current test failed and reset it to pass afterwards so this as a workaround

Flamefire replied:
Is this ok or any open question? If not please resolve so I don't loose track of open issues ;)

mat007 commented on doc/example/limitations_const_parameter_warning.cpp:
```diff
@@ -48,3 +47,10 @@ namespace limitations_const_parameter_warning_solution
+{
+    mock_base b;
+    MOCK_EXPECT(b.method).once().with(1);
+    static_cast<base*>(&b)->method(1);
```
Does this need a `static_cast`? Also it could probably be a reference?

Flamefire replied:
Yes it does. You usually pass the mocked object as a pointer to base into a method to test. Without the cast the virtual dispatch won't happen as the derived (mock) object instance is used. But for real world testing we need to test it under real world conditions. Does this make sense?

mat007 replied:
> You usually pass the mocked object as a pointer to base into a method to test

Do you think people still pass naked pointers around much these days? More than references?
(genuine question, I haven’t done much C++ in the past 2 years)

Flamefire replied:
Yes usually this would be a `unique_ptr` or similar or a reference. Most likely it would be some kind of pointer though, so that's why I'm casting to a pointer. I'll change to a lambda to avoid the explicit cast, ok?

mat007 replied:
What would it look like with a lambda?

Flamefire replied:
Changed:
```
    // Example user code taking a base* (or smart pointer variant)
    auto callMethod = [](base* bPtr){ bPtr->method(1); };
    callMethod(&b);
```

mat007 commented on doc/example/limitations_protected_private_method.cpp:
```diff
@@ -6,15 +6,16 @@
     class base
     {
+    public:
+        void call(){ method_1(); method_2(); }
```
Code formatting?

mat007 commented on doc/example/motivation.cpp:
```diff
@@ -22,6 +21,8 @@ class calculator
 };
 //]
 
+int calculator::add( int a, int b ){ return a + b; }
```
Code formatting?

mat007 commented on doc/example/patterns_async_call.cpp:
```diff
@@ -25,14 +26,23 @@ namespace
 
+namespace mock_test
+{
+    my_class::my_class( base_class& b): b(b){}
```
Code formatting?

mat007 commented on doc/example/patterns_invoke_functor.cpp:
```diff
@@ -38,7 +46,8 @@ namespace
-    MOCK_EXPECT( mock.method ).calls( boost::bind( boost::apply< void >(), _1, 42 ) ); // whenever 'method' is called, invoke the functor with 42
+    MOCK_EXPECT( mock.method ).calls( [](const auto &functor){ functor(42); } ); // whenever 'method' is called, invoke the functor with 42
     function( mock );
+    BOOST_CHECK(receivedValue == 42);
```
This shows up in the docs but the definition of `receivedValue` doesn’t.

Flamefire replied:
I wanted to keep the exposed code minimal. How shall we do this? Add a comment above this that as an example `function` calls `c.method` with a functor that sets "receivedValue" to the argument passed?

mat007 replied:
Couldn’t we move the check out of the test and into a test fixture destructor? Or maybe it’s too late there for failures to happen?

Flamefire replied:
The test fixture destructor can't assert (terminate would be called), upgrade to Boost 1.59 or 1.65 is required. (see https://github.com/mat007/turtle/pull/88#discussion_r494589023)

I added a comment

mat007 replied:
Isn’t there a way to mark the test as failed without throwing an exception?

Or 1.65, that’s fine as well 😁

mat007 commented on doc/example/patterns_retrieve_cref.cpp:
```diff
@@ -25,12 +26,22 @@ namespace
 
+namespace mock_test
+{
+    my_class::my_class( base_class& b): b(b){}
```
Formatting?

mat007 commented on doc/example/patterns_retrieve_cref.cpp:
```diff
@@ -17,6 +17,7 @@ namespace
 
     class my_class
     {
+        base_class& b;
```
It’s a bit sad that this shows up in the docs given it’s an implementation detail of the test and it’s not needed to explain the «pattern».

Flamefire replied:
On the other hand it shows users what is done with the ctor argument. I could cheat and make it a global pointer to get it out of sight.

mat007 replied:
I think we should really try to keep the code the shows up in the documentation minimal.
I know when I read code in docs I like it to be simple and straight-forward, I’m sure we all do given the choice :)

Flamefire replied:
Ok, removed that

mat007 commented on doc/example/patterns_retrieve_cref.cpp:
```diff
@@ -40,11 +51,12 @@ namespace
 
 BOOST_AUTO_TEST_CASE( method_is_called_two_times_with_the_same_value )
 {
+    using namespace mock_test;
```
Do we need this? It’s maybe better if we can keep all the code examples minimal and they all look the same.
It seems there is sometimes a namespace and sometimes not.

Flamefire replied:
I moved the test into the namespace so this is gone now

mat007 commented on doc/example/patterns_static_objects.cpp:
```diff
@@ -6,18 +6,19 @@
     {
-        my_class( int i )
-            : i_( i )
+        my_class( int i ) : i_( i )
```
OK, we can change the code formatting, I don’t really care which to choose, but this needs to be consistent everywhere 😁

mat007 commented on doc/example/patterns_static_objects.cpp:
```diff
@@ -28,13 +29,14 @@ namespace
 {
     my_class c( 42 );
     MOCK_EXPECT( f ).once().with( &c ); // the set expectation will also outlive the test case and leak into other test cases using 'f'
+    ASSERT_VERIFY_FAIL(); // Check that mock::verify fails, but that means other test fail too
```
Again this line will show up in the docs but probably make little sense for the user and may confuse her?

Flamefire replied:
See above. I wanted to a) in the docs show when and what is expected and b) test this

Flamefire replied:
I got really no better idea. Could use a fixture but that can be more confusing as the solution is to use a fixture.

mat007 commented on doc/example/rationale.cpp:
```diff
@@ -9,6 +9,19 @@
+#include <limits>
+#include <stdexcept>
+
+// Dummy to detect if the assertion unexpectedly succeeded to test what is explained
+#undef BOOST_CHECK_THROW
+#define BOOST_CHECK_THROW(expr, exc) \
```
Maybe we could collect all this workarounds to be able to test the examples/patterns in one location and use them consistently from everywhere?

Flamefire replied:
Not sure. Each is only used once so I wanted to keep them very local as they are mostly hacks to be able to test the tests. Splitting them into other files might make it even harder to understand.

mat007 commented on doc/example/reference.cpp:
```diff
@@ -403,6 +398,7 @@ namespace functor_example_1
 BOOST_AUTO_TEST_CASE( demonstrates_instantiating_a_mock_functor )
 {
    MOCK_FUNCTOR( f, void( int ) );
+   MOCK_EXPECT(f).once().with(3);
```
`MOCK_EXPECT( f ).once().with( 3 )` probably for consistency.

mat007 commented on doc/example/reference.cpp:
```diff
@@ -441,6 +438,9 @@ BOOST_AUTO_TEST_CASE( demonstrates_configuring_mock_objects )
     MOCK_EXPECT( c.method ).once().with( 0 ).in( s ).returns( 42 );
     MOCK_EXPECT( c.method2 ).never().with( "ok", mock::any );
     MOCK_EXPECT( c.method2 ).at_least( 2 ).in( s ).throws( std::runtime_error( "error !" ) );
+    BOOST_CHECK(c.method(0) == 42);
```
Should this be a `BOOST_CHECK_EQUAL`?

Flamefire replied:
Oh yes. I meant to use BOOST_TEST the universal version of this which does statement splitting (and shows both sides of the == on failure)

mat007 commented on doc/example/reference.cpp:
```diff
@@ -441,6 +438,9 @@ BOOST_AUTO_TEST_CASE( demonstrates_configuring_mock_objects )
     MOCK_EXPECT( c.method2 ).never().with( "ok", mock::any );
     MOCK_EXPECT( c.method2 ).at_least( 2 ).in( s ).throws( std::runtime_error( "error !" ) );
+    BOOST_CHECK(c.method(0) == 42);
+    BOOST_CHECK_THROW(c.method("notok", 1.f), std::runtime_error);
```
nit: `"not ok"` maybe?

mat007 commented on doc/example/reference.cpp:
```diff
@@ -379,10 +373,11 @@ MOCK_CLASS( mock_class )
 
 BOOST_AUTO_TEST_CASE( demonstrates_instantiating_a_mock_function )
 {
+    MOCK_EXPECT(f).once().with(3);
```
It’s a bit too bad that we can’t keep the introductory docs as minimal as possible, I think.

Flamefire replied:
Sure it is a little bit more but at least for me it is now easier to understand as you got some more context. And it actually tests what it says it would do

mat007 replied:
Right, that one is probably fine.

mat007 commented on doc/example/reference.cpp:
```diff
@@ -490,14 +497,17 @@ namespace invocation_example_4
 //[ invocation_example_4
 MOCK_CLASS( mock_class )
 {
-    MOCK_STATIC_METHOD( method, 1, void( int ) )
+    MOCK_STATIC_METHOD( method1, 1, void( int ) )
+    MOCK_STATIC_METHOD( method2, 1, void( int ) )
```
Why two examples for this one in particular?

Flamefire replied:
Because each gets one assertion added. I found that cleaner but could probably also add both to one function

mat007 commented on include/turtle/constraint.hpp:
```diff
@@ -136,7 +135,7 @@ namespace detail
             bool operator()( const Actual& actual ) const \
             { \
-                return Expr; \
+                (void) actual; return Expr; \
```
We could also remove the argument
```
bool operator()( const Actual& ) const \
```

Flamefire replied:
But Expr could want to use actual. And it often does. It is only sometimes that the argument is unused which creates the warning

mat007 commented on include/turtle/detail/action.hpp:
```diff
@@ -27,15 +22,19 @@ namespace detail
+        action_base& operator=(const action_base&) = delete;
+        action_base& operator=(action_base&&) = delete;
     public:
+
```
Move the blank line before `public:`?

mat007 commented on include/turtle/detail/action.hpp:
```diff
@@ -96,77 +84,54 @@ namespace detail
-                    &move< typename boost::remove_reference< Value >::type >,
-                    boost::ref( store( boost::move( v ) ) ) ) );
+            auto vRef = std::ref( store( std::move( v ) ) );
+            this->set([vRef](){ return std::move(vRef.get()); });
```
Missing some spaces.

mat007 commented on include/turtle/detail/action.hpp:
```diff
@@ -96,77 +84,54 @@ namespace detail
-                boost::bind(
-                    &move< typename boost::remove_reference< Value >::type >,
-                    boost::ref( store( boost::move( v ) ) ) ) );
+            auto vRef = std::ref( store( std::move( v ) ) );
```
`vRef` -> `vref`?

Flamefire replied:
Wanted to camel case this. It's a ref of v

mat007 commented on include/turtle/detail/formatter.hpp:
```diff
@@ -11,7 +11,7 @@
 #include "../config.hpp"
 #include "../stream.hpp"
-#include "addressof.hpp"
+#include <memory.h>
```
Shouldn’t this be `<memory>`?

mat007 commented on include/turtle/detail/is_functor.hpp:
```diff
@@ -10,51 +10,19 @@
-            has_result< T >,
-            has_sig< T >
-        >
+    /// Trait to return true if F is a functor that can be called with a single argument Arg
```
nit: `///` -> `//`?

Flamefire replied:
Triple slash is often understood es docu instead of a comment by IDEs
So I'd like to keep them

mat007 replied:
Right, sure why not.

mat007 commented on include/turtle/detail/type_name.hpp:
```diff
@@ -51,15 +40,17 @@ namespace detail
-                &std::free );
+            struct Deleter
+            {
+                void operator()(const void* p) { std::free(const_cast<void*>(p)); }
```
Missing spaces?

mat007 commented on include/turtle/detail/void_t.hpp:
```diff
@@ -0,0 +1,27 @@
+    {
+        using type = void;
+    };
+    /// Standard helper to implement the detection idiom. Returns always void
```
nit: `///` -> `//`?

mat007 commented on include/turtle/mock.hpp:
```diff
@@ -21,133 +21,126 @@
-#include <boost/utility/identity_type.hpp>
-#include <boost/mpl/assert.hpp>
 
+/// MOCK_CLASS( name )
+/// Define a class
```
Triple slash comments again, is this a new thing I don’t know about? 😀

mat007 commented on test/detail/test_type_name.cpp:
```diff
@@ -7,15 +7,17 @@
     {
-        return boost::lexical_cast< std::string >( MOCK_TYPE_NAME(T) );
+        std::ostringstream s;
+        s << mock::detail::make_type_name(t); // Typename can be streamed
```
`Typename` -> `T`?

Flamefire replied:
No I actually meant "typename" the name of the type of `t` which is returned by `make_type_name`

mat007 replied:
Right, so maybe `// name of the type of T can be streamed`?

Flamefire replied:
I just removed the comment. I mean: It is streamed so obviously it can be streamed ;D

mat007 commented on test/test_log.cpp:
```diff
@@ -54,13 +54,19 @@ BOOST_AUTO_TEST_CASE( pointer_yields_its_value_when_serialized )
         int i = 0;
+        std::ostringstream s;
+        s << &i;
+        const std::string pointerValue = s.str();
```
nit: could probably be `const auto`?

Flamefire replied:
While I do like `auto` I tend to use it only when the type appears on the right hand side. E.g. `auto p = static_cast<int>(u)`. I find that more readable in the end. This case could be considered a corner case as `str()` is almost `string` but I think it is still better to explicitly see that `pointerValue` is a string.

Flamefire commented on doc/example/customization.cpp:
```diff
@@ -86,8 +87,7 @@ struct near_constraint
     {
-        return std::abs( actual - boost::unwrap_ref( expected_ ) )
-            < boost::unwrap_ref( threshold_ );
+        return std::abs( actual - expected_ ) < threshold_ ;
```
Tbh I don't understand that part of the docs. Imo the important part is to use (c) ref to pass things as references instead of values. The use and even existence of unwrap_ref is not required and I'd rather remove it from the docs. A cref is simply a copy able reference

Flamefire replied:
It seems it is still required. Just not here. Example: Comparing an actual `const char*` against an expected `cref(std::string)` requires converting both sides: the cref to a string and the pointer to a string. This doesn't work (only 1 conversion allowed). However here it is only int vs cref(int) which works fine. even int vs cref(float) works.

Added a new test case to showcase that

I added it back here.

mat007 replied:
Nice, thanks!

Flamefire commented:
As many comments are for code style: would you mind creating a clangformat file for clangformat 9 which I could use? I find it pretty hard to guess how to do that otherwise... Or if you don't care which style is used I could use one from another project. Avoids the bikeshedding and confusion... 

About the docs: sure I can extract that but this would be hard as I had to fix them and used the new changes already there. So they would need to be merged afterwards. But then I feel that test coverage is missing. Some macros are only used there and I might miss a bug. 
Maybe leave them in here until the pr is ready and I extract them to a follow up just before merging if you really want them separately. 

I'll address the other comments next week when I work on this again. A style guide or definition until then would be nice. Check https://github.com/Return-To-The-Roots or https://github.com/boostorg/nowide for example configs which would be easiest

Flamefire commented:
I rebased this to current master to resolve the conflict.

@mat007 I'm still waiting for your call on using clang-format which would greatly reduce the work for keeping the formatting uniform.

Flamefire commented:
@mat007 Kindly requesting on guidance how to continue on this. If wanted we could meet in a chat or so if there is stuff for discussion so it is faster than through github. I'd like to build more stuff on top of this like replacing the preprocessor generated implementations by variadic templates to make it easier to read and debug and using a macro that allows method specifiers (const, ref qualifiers) and especially override support. This will make the mock much safer and avoids many warnings we have in our code.

mat007 commented:
> @mat007 Kindly requesting on guidance how to continue on this.

Sorry, I’ve been away quite a lot over the summer.
I think they were two remaining points that we need solving: formatting and docs.

> As many comments are for code style: would you mind creating a clangformat file for clangformat 9 which I could use? I find it pretty hard to guess how to do that otherwise... Or if you don't care which style is used I could use one from another project. Avoids the bikeshedding and confusion...

I don’t care as long as it’s used everywhere in the code.
Doesn’t Boost suggest a `clangformat` file? Or some of the library authors?

> About the docs: sure I can extract that but this would be hard as I had to fix them and used the new changes already there. So they would need to be merged afterwards. But then I feel that test coverage is missing. Some macros are only used there and I might miss a bug.
Maybe leave them in here until the pr is ready and I extract them to a follow up just before merging if you really want them separately.

I liked how the reference documentation only explained each bit one at a time, but I can understand how making sure it compiles helps.
I’ll take a look to see what it looks like now, but I don’t think it’s a blocker for the PR as it can be adjusted later if needed.

Again, sorry for my poor reactivity this last couple of months!

Flamefire commented:
Opened #97 to discuss the style first before applying here.

Re docs: Yes please have a look, maybe we can reduce the range of what we show. I already tried to keep as much of the visual part as possible, hence the define trick with the exception throw check and others.

Flamefire commented:
@mat007 As #97 is now sorted out I'd like you to review this excluding style issues. Reformatting the files will introduce so many changes that reviewing details will become unfeasible. In fact I think we should even merge this first and then open another PR with the reformatting on top. This allows to focus on features here and formatting issues that may need to be resolved separately. Or what do you say?

mat007 commented on doc/example/limitations_non_virtual_method.cpp:
```diff
@@ -25,3 +24,10 @@ MOCK_BASE_CLASS( mock_base, base )
+{
+    mock_base b;
+    MOCK_EXPECT(b.method).never();
+    static_cast<base*>(&b)->method(); // Doesn't call the mocked method
```
nit: I think the comment is not needed given the automatic documentation provided by the previous line 😁

Flamefire replied:
I extended the comment a bit. I think it is still important as it should show a very specific point

mat007 commented on doc/example/limitations_template_method.cpp:
```diff
@@ -36,6 +34,14 @@ MOCK_CLASS( mock_concept )
+{
+    mock_concept b;
+    MOCK_EXPECT(b.method_int).once().with(42);
+    MOCK_EXPECT(b.method_string).once().with(mock::equal(std::string("string")));
```
`mock::equal("string")` should work, right?

Flamefire replied:
The method takes a `const char*` so `mock::equal("string")` would compare pointers, not the values. The conversion to string makes it compare the actual strings instead

mat007 replied:
Oh yeah, there used to be an overload for `mock::equal` with a `const char*` to act as a special case to compare strings, but this was removed in a later version of turtle 😬 
Maybe
```suggestion
    MOCK_EXPECT(b.method_string).once().with(mock::equal<std::string>("string"));
```
though?

Flamefire replied:
I don't see any advantage here. Passing a string is IMO more clear than changing template params

mat007 replied:
I still think `mock::equal<std::string>` is a bit clearer, as in meaning «equal as a string», but that’s pretty minor.

mat007 commented on doc/example/limitations_template_method.cpp:
```diff
@@ -80,4 +86,12 @@ std::string mock_concept::create< std::string >()
+{
+    mock_concept b;
+    MOCK_EXPECT(b.create_int).once().returns(int{});
+    MOCK_EXPECT(b.create_string).once().returns(std::string{});
```
`returns("")` should work?

Flamefire replied:
Yes

mat007 commented on doc/example/limitations_template_method.cpp:
```diff
@@ -80,4 +86,12 @@ std::string mock_concept::create< std::string >()
+BOOST_AUTO_TEST_CASE(dispatch_methods_are_called)
+{
+    mock_concept b;
+    MOCK_EXPECT(b.create_int).once().returns(int{});
```
Why not `returns(0)`?

mat007 commented on doc/example/patterns_async_call.cpp:
```diff
@@ -25,14 +26,23 @@ namespace
+    my_class::my_class( base_class& b): b(b){}
+    void my_class::flush()
+    {
+        static int secret_value = 7;
```
Why the secret? 😀 Can’t it just be `count` or something like that?

Flamefire replied:
Good point

mat007 commented on doc/example/patterns_invoke_functor.cpp:
```diff
@@ -7,24 +7,32 @@
+namespace
+{
+    int receivedValue = 0;
+    void setx(int newValue)
```
Why the `x`?

Flamefire replied:
`setValue` better?

mat007 replied:
Yeah or maybe even `set`, either way.

mat007 commented on doc/example/reference.cpp:
```diff
@@ -490,14 +497,17 @@ namespace invocation_example_4
-    MOCK_EXPECT( c.method ).once();
-    MOCK_EXPECT( mock_class::method ).once(); // does the same
+    MOCK_EXPECT( c.method1 ).once();
+    MOCK_EXPECT( mock_class::method2 ).once(); // does the same
```
Well, it doesn’t do the same anymore if they’re on different methods.

Flamefire replied:
Adapted the comment, better?

mat007 replied:
OK.

mat007 commented on doc/example/reference.cpp:
```diff
@@ -507,14 +517,17 @@ namespace constraints_example_1
-   MOCK_EXPECT( c.method ).with( mock::equal( 3 ), mock::equal( "some string" ) );
-   MOCK_EXPECT( c.method ).with( 3, "some string" );                               // equivalent to the previous one using short-cuts
+   MOCK_EXPECT( c.method1 ).with( mock::equal( 3 ), mock::equal( "some string" ) );
+   MOCK_EXPECT( c.method2 ).with( 3, "some string" );                               // equivalent to the previous one using short-cuts
```
It’s not exactly equivalent anymore now if it’s two different methods.

Flamefire replied:
Adapted the comment, better?

mat007 replied:
OK.

mat007 commented on test/test_constraint.cpp:
```diff
@@ -25,8 +23,6 @@ BOOST_AUTO_TEST_CASE( mock_constraint_is_supported_by_compilers_with_variadic_ma
-
 namespace
 {
     MOCK_CONSTRAINT_EXT( constraint_0_ext, 0,, actual == 0 )
```
Do we still need `MOCK_CONSTRAINT_EXT`?

Flamefire replied:
You are right, we don't. It is an implementation detail now

mat007 commented on include/turtle/detail/function_template.hpp:
```diff
@@ -9,7 +9,7 @@
 
 #define MOCK_MOVE(z, n, d) \
-    mock::detail::move_if_not_lvalue_reference< T##n >( t##n )
+    std::forward< T##n >( t##n )
```
nit: this could probably be renamed `MOCK_FORWARD`

mat007 commented on include/turtle/mock.hpp:
```diff
@@ -21,133 +21,126 @@
+    mock::detail::functor< MOCK_FUNCTION_TYPE(__VA_ARGS__) > f, f##_mock
+/// MOCK_FUNCTOR_TPL( name, signature )
+/// Deprecated. Same as MOCK_FUNCTOR
+#define MOCK_FUNCTOR_TPL(f, ...) MOCK_FUNCTOR(f, __VA_ARGS__)
```
Should we not just remove them? I don’t mind breaking users code in a new release if we give them a simple solution to fix it quickly.
Or maybe actually we should keep the deprecated macros but make it trigger a compilation error explaining how to change the code?

Flamefire replied:
I can expand it to `static_assert(false, "MOCK_FUNCTOR_TPL has been replaced by MOCK_FUNCTOR")` :)

mat007 replied:
Nice 😁

mat007 commented on test/detail/test_function.cpp:
```diff
@@ -546,6 +535,19 @@ BOOST_FIXTURE_TEST_CASE( triggering_an_expectation_returns_the_set_auto_ptr_valu
         CHECK_CALLS( 2 );
     }
+    {
+        mock::detail::function< std::auto_ptr< int >() > f;
```
Should we bother maintaining `auto_ptr` support?

Flamefire replied:
Good point. 150 lines less code to maintain

Flamefire commented on doc/example/customization.cpp:
```diff
@@ -141,6 +141,11 @@ BOOST_AUTO_TEST_CASE( forty_one_plus_one_is_forty_two_plus_or_minus_one )
 #include <turtle/mock.hpp>
 //]
 
+#if defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#endif
```
Unfortunately not. The problem here is that full signatures should be shown but the parameters are not used. One solution is to `(void) param` them each or to remove the name. Both is not good for explaining this code, So this is IMO already the best solution

Flamefire commented on include/turtle/matcher.hpp:
```diff
@@ -28,13 +27,11 @@ namespace mock
         {
-            return mock::equal(
-                boost::unwrap_ref( expected_ ) ).c_( actual );
+            return mock::equal( mock::unwrap_ref( expected_ ) ).c_( actual );
```
It is required in general. But I'm actually not sure if it is required here as `equal` can already unwrap the ref wrapper. I'll leave it for now

mat007 replied:
Fair enough

Flamefire commented:
Friendly remainder that the review comments have been addressed so far. :)

mat007 commented:
> Friendly remainder that the review comments have been addressed so far. :)

Sorry, I’ve been meaning to do it, I swear!
Thanks for the nudge 😄

jonesmz commented:
I notice that this PR has a failing test case.

Is that failure intentional, or part of the original code, and not caused by this PR ?

Flamefire commented:
> I notice that this PR has a failing test case.
> 
> Is that failure intentional, or part of the original code, and not caused by this PR ?

Haven't seen there were updates to this, hence ignored that pending replies from @mat007 

Incorporated the questions/suggestions, fixed the failure and updated CI

FTR: The failing check was caused by `near` comparing against "less-than tolerance" which makes "0" as the tolerance impossible. Changed that.

Flamefire commented:
@mat007 After some heavy fighting against the CI and some peculiarities of the Ubuntu linker I finally managed to get this working on CI again including the coverage report.
I addressed your review comments unfortunately the history got a bit messed up. Basically the last 8 commits starting at https://github.com/mat007/turtle/pull/88/commits/728dfd06eb82d063b6b7df82de94405e5b98a869 are new incorporating your requests and the new CI. The tests now also pass on C++20.

jonesmz commented on include/turtle/detail/parameter.hpp:
```diff
@@ -10,24 +10,70 @@
+        using type = U;
     };
+    template<typename T>
+    using parameter_type_t = typename parameter_type<T>::type;
```
Would it make sense to use boost::callable_traits for parameter type deduction instead?

Flamefire replied:
Would be possible. But similar to `is_callable` this only supports simple methods which seemingly is(/was?) enough, so this (much smaller) implementation should suffice.
And Boost.CallableTraits was introduced in Boost 1.66, i.e. it would need another increase in the minimum Boost version required which I wanted to avoid at the time of writing this.

jonesmz commented on include/turtle/log.hpp:
```diff
@@ -51,16 +50,17 @@ namespace detail
+    {};
+    template<typename T>
+    struct is_callable: is_callable_impl< std::remove_cv_t<T> >
+    {};
```
https://www.boost.org/doc/libs/master/libs/callable_traits/doc/html/callable_traits/reference.html#callable_traits.reference.ref_is_invocable ?

Flamefire replied:
Not really. I tried to translate what was there before and this whole thing isn't really good anyway: Besides a plain function there isn't a Type "callable" (this trait only takes a single parameter!)
Actually you would need to check if you can invoke something with specific parameters as `is_invocable` does. But we don't have those.

So a better name would be e.g. `is_function`.

mat007 reviewed (approved):
Thanks @Flamefire this is a huge achievement, and I have to tip my hat off to you for commitment! 😁 

I can either merge the PR, or I can add you as a collaborator to the project and you can merge it yourself? (you don’t have to accept if you don’t feel like it, no worries)

Flamefire commented:
> I can either merge the PR, or I can add you as a collaborator to the project and you can merge it yourself? (you don’t have to accept if you don’t feel like it, no worries)

Either is fine for me ;)
Next thing I'll open a PR for is the formatting change as previewed in #97
